vllm/model_executor/models/fuyu.py

```diff
@@ -0,0 +1,202 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch Fuyu model."""
+from typing import List, Optional
+
+import torch
+import torch.utils.checkpoint
+from torch import nn
+from transformers import FuyuConfig
+
+from vllm.attention import AttentionMetadata
+from vllm.config import VisionLanguageConfig
+from vllm.model_executor.layers.linear import (LinearMethodBase,
+                                               RowParallelLinear)
+from vllm.model_executor.models.persimmon import PersimmonForCausalLM
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.model_executor.weight_utils import (default_weight_loader,
+                                              hf_model_weights_iterator)
+from vllm.sequence import SamplerOutput
+
+
+class FuyuInputProcessor:
+    """Simple Processor for padding and patch PIXEL_VALUES image input"""
+
+    def __init__(self, patch_size=30, bos_token_id=1) -> None:
+        self.patch_size = patch_size
+        self.bos_token_id = bos_token_id
+
+    def _calculate_patch_nums(self, x: int) -> int:
+        x = x // self.patch_size
+        if x % self.patch_size != 0:
+            x += 1
+        return x
+
+    def _patch_image(self, image):
+        patches = list(image.split(self.patch_size, dim=0))
+        patches = [
+            torch.stack(patch.split(self.patch_size, dim=1))
+            for patch in patches
+        ]
+        patches = torch.concat(patches, dim=0)
+        return patches
+
+    def __call__(self, image: torch.Tensor) -> torch.Tensor:
+        batch_size, num_channels, H, W = image.shape
+        assert batch_size == 1
+        assert num_channels == 3, "Input image should have 3 channels"
+
+        image = image.squeeze(0).permute(1, 2, 0)
+        patch_y = self._calculate_patch_nums(H)
+        patch_x = self._calculate_patch_nums(W)
+
+        image_padded = torch.ones(
+            patch_y * self.patch_size,
+            patch_x * self.patch_size,
+            num_channels,
+            dtype=image.dtype,
+            device=image.device,
+        )
+        image_padded[:H, :W, :] = image
+        image_patches = self._patch_image(image_padded).view(
+            patch_y * patch_x, self.patch_size**2 * num_channels)
+        return image_patches
+
+
+class FuyuForCausalLM(nn.Module):
+
+    def __init__(
+        self,
+        config: FuyuConfig,
+        vision_language_config: VisionLanguageConfig,
```

Comment:
since we are not using  image_input_shape and image_feature_size in this model I wonder if we want to make these optional in the VisionLanguageConfig and not set it when initializing the LLM(....)?

Reply:
It seems that we can still pass `MultiModalData` in generation without passing `vision_language_config` in LLM(....) initialization. 

So maybe we can hardcoded `image_token_id` in `FuyuForCausalLM`, then initialize LLM like other decoder-only model and generate outputs like other vision model:
```python
    llm = LLM(model="adept/fuyu-8b")

    # load and create image prompt
    images = torch.load("images/cherry_blossom_pixel_values.pt")

    _, _, H, W = images.shape
    nrow = H // 30 + 1 if H % 30 else H // 30
    ncol = W // 30 + 1 if W % 30 else W // 30

    image_prompt = ("|SPEAKER|" * ncol + "|NEWLINE|") * nrow
    prompt = image_prompt + "Generate a coco-style caption.\n"

    # inference
    outputs = llm.generate(
        prompt,
        multi_modal_data=MultiModalData(MultiModalData.Type.IMAGE, images),
    )
```

In my opinion, most of vision LLM has encoder-decoder architecture and `Fuyu-8B` is a decoder-only exception. Making `image_input_shape` and `image_feature_size` optional may be not a good choice, because this will bypass the config checking and cause some unexpected errors for other vision models.

Reply:
Well, after investigating the `model_runner.py`, I think making `image_input_shape` and `image_feature_size` optional would be a better choice than I have mentioned above.

But I think warnings should be raised for users if they passed an incomplete `vision_language_config`.

Reply:
Wonder if we want to add an assert here https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llava.py#L70 for input_image_shape and image_feature_size size for llava since we made these optional.

Reply:
OK. I have expanded the `vision_language_config` assertion to cover `input_image_shape` and `image_feature_size`.